Cabal/src/Distribution/Simple/Compiler.hs

```diff
@@ -296,20 +309,32 @@ data OptimisationLevel
 instance Binary OptimisationLevel
 instance Structured OptimisationLevel
 
+instance Parsec OptimisationLevel where
+  parsec = parsecOptimisationLevel
+
+parsecOptimisationLevel :: CabalParsing m => m OptimisationLevel
+parsecOptimisationLevel = boolParser <|> intParser
+  where
+    boolParser = (bool NoOptimisation NormalOptimisation) <$> parsec
+    intParser = intToOptimisationLevel <$> integral
+
 flagToOptimisationLevel :: Maybe String -> OptimisationLevel
 flagToOptimisationLevel Nothing = NormalOptimisation
 flagToOptimisationLevel (Just s) = case reads s of
-  [(i, "")]
-    | i >= fromEnum (minBound :: OptimisationLevel)
-        && i <= fromEnum (maxBound :: OptimisationLevel) ->
-        toEnum i
-    | otherwise ->
-        error $
-          "Bad optimisation level: "
-            ++ show i
-            ++ ". Valid values are 0..2"
+  [(i, "")] -> intToOptimisationLevel i
   _ -> error $ "Can't parse optimisation level " ++ s
 
+intToOptimisationLevel :: Int -> OptimisationLevel
+intToOptimisationLevel i
+  | i >= fromEnum (minBound :: OptimisationLevel)
+      && i <= fromEnum (maxBound :: OptimisationLevel) =
+      toEnum i
+  | otherwise =
+      error $
+        "Bad optimisation level: "
+          ++ show i
+          ++ ". Valid values are 0..2"
```

Comment:
maybe the message should also use minBound/maxBound instead of hard-coded values 0 and 2?

Reply:
Agreed, that makes you future-proof against changes to the optimization levels.

Reply:
Should be fixed now: https://github.com/haskell/cabal/pull/8889/files#diff-1b1d4d646bf0cab8578f54a6a838a342c7a5787ea7469e2cf195a8959fc5a122R328